File: harness-bundle/src/shared-trash-client.ts
import { request as httpRequest } from "node:http";
import { platformQuotaConfiguration } from "./quota-client.js";

export const SHARED_TRASH_REQUEST_TIMEOUT_MS = 180_000;
export const SHARED_TRASH_TOOL_TIMEOUT_MS = 210_000;

export type SharedTrashOperation = {
  projectId: string;
  operation: "list" | "recycle" | "restore";
  path?: string;
  entryId?: string;
};

export class SharedTrashError extends Error {
  constructor(
    readonly status: number,
    code: string,
  ) {
    super(code);
  }
}

/** Uses the employee's existing authenticated platform channel. */
export class PlatformSharedTrashClient {
  constructor(
    readonly configuration: NonNullable<
      ReturnType<typeof platformQuotaConfiguration>
    >,
  ) {}

  static fromEnvironment(environment: NodeJS.ProcessEnv = process.env) {
    const configuration = platformQuotaConfiguration(environment);
    return configuration && new PlatformSharedTrashClient(configuration);
  }

  async operate(
    input: SharedTrashOperation,
    source?: "workspace-store",
  ): Promise<unknown> {
    const url = new URL(
      "internal/runtime/shared-trash",
      this.configuration.baseURL,
    );
    const headers = {
      authorization: `Bearer ${this.configuration.token}`,
      "content-type": "application/json",
    };
    const body = JSON.stringify({
      ...input,
      ...(source ? { source } : {}),
      sid: this.configuration.sid,
    });
    // Once the owner's Store admits a deletion, it must keep its move lock
    // until the central transaction acknowledges it, even if the browser or
    // calling tool disconnects. Node HTTP avoids fetch's implicit header timer.
    const response = source
      ? await new Promise<Response>((resolve, reject) => {
          const request = httpRequest(
            url,
            { method: "POST", headers, agent: false },
            (response) => {
              void (async () => {
                const chunks: Buffer[] = [];
                for await (const chunk of response)
                  chunks.push(Buffer.from(chunk));
                resolve(
                  new Response(Buffer.concat(chunks), {
                    status: response.statusCode!,
                  }),
                );
              })().catch(reject);
            },
          );
          request.on("error", reject);
          request.end(body);
        })
      : await fetch(url, {
          method: "POST",
          headers,
          body,
          signal: AbortSignal.timeout(SHARED_TRASH_REQUEST_TIMEOUT_MS),
        });
    if (!response.ok) {
      let code = `shared_trash_http_${response.status}`;
      try {
        const payload = (await response.json()) as { error?: unknown };
        if (typeof payload.error === "string") code = payload.error;
      } catch {
        // Preserve the status if the upstream failure has no JSON body.
      }
      throw new SharedTrashError(response.status, code);
    }
    return response.json();
  }
}

File: internal/userhost/shared_trash_client.go
package userhost

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// Reuse the existing employee platform identity. Unlike best-effort audit,
// recycling must succeed centrally before the file operation reports success.
func (c *auditClient) recycleSharedFile(ctx context.Context, projectID, path string) error {
	body, _ := json.Marshal(map[string]string{
		"sid": c.sid, "projectId": projectID, "operation": "recycle", "path": path,
	})
	endpoint := strings.TrimSuffix(c.endpoint, "/audit") + "/shared-trash"
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(body))
	if err != nil {
		return err
	}
	request.Header.Set("Authorization", "Bearer "+c.credential)
	request.Header.Set("Content-Type", "application/json")
	client := *c.client
	client.Timeout = 3 * time.Minute
	response, err := client.Do(request)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		var payload struct {
			Error string `json:"error"`
		}
		if json.NewDecoder(io.LimitReader(response.Body, 4096)).Decode(&payload) == nil && payload.Error != "" {
			return errors.New(payload.Error)
		}
		return fmt.Errorf("shared_trash_http_%d", response.StatusCode)
	}
	return nil
}
